File: list/Solution92.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution92 {
    public ListNode reverseBetween(ListNode head, int m, int n) {
    	ListNode sentinel = new ListNode();
    	sentinel.next = head;
    	int length = n - m + 1;
    	ListNode point1 = sentinel;
    	while(m > 1){
    		point1 = point1.next;
    		m -= 1;
    	}
    	ListNode point2 = point1.next;
    	ListNode[] pointers = new ListNode[length];
    	int index = 0;
    	while(index < length){
    		pointers[index] = point2;
    		point2 = point2.next;
    		index += 1;
    	}
    	pointers[0].next = point2;
    	for(int i = 1; i < length; i++){
    		pointers[i].next = pointers[i - 1];
    	}
    	point1.next = pointers[length - 1];
    	return sentinel.next;
    }

    public static void main(String[] args){
    	Solution92 s = new Solution92();
    	ListNode test = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    	ListNode result = s.reverseBetween(test,2, 4);
    }
}

File: C++/Solution92.cpp
#include<iostream>
#include<vector>

#include<unordered_map>

using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution92 {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode* sentinel = new ListNode(0, head);
        ListNode* node1 = sentinel;
        int left2 = left;
        while(left > 1){
            left--;
            node1 = node1 -> next;
        }
        vector<ListNode*> vec(right - left + 1);
        ListNode* node2 = node1;
        for(int i = 0; i < right - left2 + 1; i++){
            node2 = node2 -> next;
            vec[i] = node2;
            //cout << node2 -> val << endl;
        }
        ListNode* copy = node2 -> next;
        node1 -> next = node2;
        for(int i = right - left2; i >= 1; i--){
            vec[i] -> next = vec[i - 1];
        }
        vec[0] -> next = copy;
        return sentinel -> next;
    }
};